### PR TITLE
Allow read a file paths with upper case symbols

### PR DESCRIPTION
**Pull request by ChugunovRoman**
These changes allow the engine to read files containing uppercase characters.\
On Windows this shouldn't take effect

### CODE REVIEW

File: src/xrCore/xr_ini.cpp [hunk 1]
```diff
@@ -451,13 +451,22 @@ void CInifile::Load(IReader* F, pcstr path, allow_include_func_t allow_include_f
             R_ASSERT(path && path[0]);
             if (_GetItem(str, 1, inc_name, '"'))
             {
+#ifdef XR_PLATFORM_WINDOWS
```
Comment (Xottab-DUTY) [inline src/xrCore/xr_ini.cpp]: This is ridiculous, please do something with it. `xr_fs_nostrlwr` is already a macro...

File: src/xrCore/LocatorAPI.cpp [hunk 2]
```diff
@@ -592,16 +592,50 @@ bool CLocatorAPI::load_all_unloaded_archives()
...
+    xr_fs_strlwr(N);
 #elif defined(XR_PLATFORM_LINUX) || defined(XR_PLATFORM_FREEBSD)
     xr_strcpy(N, sizeof N, entry.name);
 #endif
-    xr_fs_strlwr(N);
+
+    bool ignore = ignore_name(N);
+    if (ignore) {
+        return;
+    }
```
Comment (Xottab-DUTY) [inline src/xrCore/LocatorAPI.cpp]: This change is not related to this pull request main goal. Please, do it as separate commit.

File: src/xrCore/xr_ini.cpp (hunk 1 above)
Comment (ChugunovRoman) [reply src/xrCore/xr_ini.cpp]: @Xottab-DUTY fixed

File: src/xrCore/LocatorAPI.cpp (hunk 2 above)
Comment (ChugunovRoman) [reply src/xrCore/LocatorAPI.cpp]: @Xottab-DUTY fixed

File: src/xrCore/xr_ini.cpp [hunk 3]
```diff
@@ -451,13 +451,20 @@ void CInifile::Load(IReader* F, pcstr path, allow_include_func_t allow_include_f
             R_ASSERT(path && path[0]);
             if (_GetItem(str, 1, inc_name, '"'))
             {
-                xr_fs_nostrlwr(inc_name); // compensate removed xr_strlwr on path on Linux, etc
-
+                xr_fs_strlwr(inc_name); // compensate removed xr_strlwr on path on Linux, etc
```
Comment (Xottab-DUTY) [inline src/xrCore/xr_ini.cpp]: That changes the behaviour on Windows!

File: src/xrCore/LocatorAPI_defs.cpp [hunk 4]
```diff
@@ -105,6 +105,17 @@ LPCSTR FS_Path::_update(string_path& dest, LPCSTR src) const
     R_ASSERT(src);
     string_path temp;
     xr_strcpy(temp, sizeof(temp), src);
+
+#ifdef XR_PLATFORM_LINUX
+    string_path fullPath;
+    strconcat(sizeof(fullPath), fullPath, m_Path, temp);
+    xr_fs_strlwr(fullPath);
+    if (FS.exist(fullPath, FSType::External) || FS.exist(fullPath, FSType::Virtual)) {
```
Comment (Xottab-DUTY) [inline src/xrCore/LocatorAPI_defs.cpp]:
```suggestion
    if (FS.exist(fullPath, FSType::External) || FS.exist(fullPath, FSType::Virtual))
    {
```

File: src/xrCore/LocatorAPI_defs.cpp [hunk 5]
```diff
@@ -105,6 +105,17 @@ LPCSTR FS_Path::_update(string_path& dest, LPCSTR src) const
     R_ASSERT(src);
     string_path temp;
     xr_strcpy(temp, sizeof(temp), src);
+
+#ifdef XR_PLATFORM_LINUX
+    string_path fullPath;
+    strconcat(sizeof(fullPath), fullPath, m_Path, temp);
```
Comment (Xottab-DUTY) [inline src/xrCore/LocatorAPI_defs.cpp]: sizeof(fullPath) not needed
```suggestion
    strconcat(fullPath, m_Path, temp);
```

File: src/xrCore/LocatorAPI_defs.cpp [hunk 6]
```diff
@@ -105,6 +105,17 @@ LPCSTR FS_Path::_update(string_path& dest, LPCSTR src) const
...
+    xr_fs_strlwr(fullPath);
```
Comment (Xottab-DUTY) [inline src/xrCore/LocatorAPI_defs.cpp]: Don't use `xr_fs` if you already use `#ifdef XR_PLATFORM_LINUX`, just use `xr_`

File: src/xrCore/LocatorAPI_defs.cpp [hunk 7]
```diff
@@ -105,6 +105,17 @@ LPCSTR FS_Path::_update(string_path& dest, LPCSTR src) const
...
+    if (FS.exist(fullPath, FSType::External) || FS.exist(fullPath, FSType::Virtual)) {
+        xr_strcpy(dest, sizeof(dest), fullPath);
```
Comment (Xottab-DUTY) [inline src/xrCore/LocatorAPI_defs.cpp]: sizeof(dest) not needed
```suggestion
        xr_strcpy(dest, fullPath);
```

File: src/xrCore/xr_ini.cpp [hunk 8]
```diff
@@ -451,13 +451,20 @@ void CInifile::Load(IReader* F, pcstr path, allow_include_func_t allow_include_f
...
                 string_path fn;
                 strconcat(sizeof fn, fn, path, inc_name);
                 if (!allow_include_func || allow_include_func(fn))
                 {
                     IReader* I = FS.r_open(fn);
+#ifdef XR_PLATFORM_LINUX
+                    if (I == nullptr)
+                    {
+                        xr_fs_nostrlwr(inc_name);
+                        strconcat(sizeof fn, fn, path, inc_name);
```
Comment (Xottab-DUTY) [inline src/xrCore/xr_ini.cpp]: `sizeof fn` not needed
```suggestion
                        strconcat(fn, path, inc_name);
```

File: src/xrCore/LocatorAPI.cpp [hunk 9]
```diff
@@ -603,6 +632,11 @@ void CLocatorAPI::ProcessOne(pcstr path, const _finddata_t& entry)
 #endif
     xr_fs_strlwr(N);
 
+    bool ignore = ignore_name(N);
+    if (ignore) {
+        return;
+    }
```
Comment (Xottab-DUTY) [inline src/xrCore/LocatorAPI.cpp]: Excessive variable, parentheses should be on the next line, but in this case they are not even needed.
The code should look like this:
```cpp
if (ignore_name(N))
    return;
```

**Comment by Xottab-DUTY** [comment]
@ChugunovRoman

File: src/xrCore/LocatorAPI_defs.cpp (hunk 4 above)
Comment (ChugunovRoman) [reply src/xrCore/LocatorAPI_defs.cpp]: Done

File: src/xrCore/LocatorAPI_defs.cpp (hunk 5 above)
Comment (ChugunovRoman) [reply src/xrCore/LocatorAPI_defs.cpp]: Done

File: src/xrCore/LocatorAPI_defs.cpp (hunk 7 above)
Comment (ChugunovRoman) [reply src/xrCore/LocatorAPI_defs.cpp]: Done

File: src/xrCore/xr_ini.cpp (hunk 8 above)
Comment (ChugunovRoman) [reply src/xrCore/xr_ini.cpp]: Done

File: src/xrCore/LocatorAPI.cpp (hunk 9 above)
Comment (ChugunovRoman) [reply src/xrCore/LocatorAPI.cpp]: Done

File: src/xrCore/xr_ini.cpp (hunk 3 above)
Comment (ChugunovRoman) [reply src/xrCore/xr_ini.cpp]: Why? Windows is not case-sensitive in paths. If you do not change the case of characters in this place, will nothing change for windows? So?

File: src/xrCore/LocatorAPI_defs.cpp (hunk 6 above)
Comment (ChugunovRoman) [reply src/xrCore/LocatorAPI_defs.cpp]: Done

File: src/xrCore/xr_ini.cpp (hunk 3 above)
Comment (Xottab-DUTY) [reply src/xrCore/xr_ini.cpp]: Because strings comparison inside the engine is case-sensitive.

Comment (ChugunovRoman) [reply src/xrCore/xr_ini.cpp]: The `xr_fs_nostrlwr` function does not change case on Windows, and `xr_fs_strlwr` does not change case on Linux. So maybe remove the `xr_fs_strlwr` function here?

Comment (ChugunovRoman) [reply src/xrCore/xr_ini.cpp]: I think call of `xr_fs_strlwr` unnecessary

Comment (ChugunovRoman) [reply src/xrCore/xr_ini.cpp]: Removed

**Review by Xottab-DUTY — APPROVED** [review]

**Comment by Xottab-DUTY** [comment]
Thank you!
